lektor/watcher.py

```diff
@@ -2,59 +2,86 @@
-            path = (
-                event.dest_path if isinstance(event, FileMovedEvent) else event.src_path
-            )
-            item = (time.time(), event.event_type, path)
-            if self.queue is not None:
-                self.queue.put(item)
+            if isinstance(event, FileMovedEvent):
+                path = event.dest_path
             else:
-                self.callback(*item)
+                path = event.src_path
+            self.queue.put((time.time(), event.event_type, path))
 
 
 class BasicWatcher:
-    def __init__(self, paths, callback=None):
-        self.event_handler = EventHandler(callback=callback)
-        self.observer = Observer()
-        for path in paths:
-            self.observer.schedule(self.event_handler, path, recursive=True)
-        self.observer.setDaemon(True)
+    def __init__(self, paths, observer_classes=(Observer, PollingObserver)):
+        self.event_handler = EventHandler()
+        self.paths = paths
+        self.observer_classes = observer_classes
+        self.observer = None
+
+    def start(self):
+        untried = set(self.observer_classes)
+        for observer_class in self.observer_classes:
+            try:
+                self._start_observer(observer_class)
+            except Exception as exc:
+                untried.discard(observer_class)
+                if len(untried) == 0:
+                    raise
+                click.secho(
+                    f"Creation of {observer_class.__module__}.{observer_class.__name__} "
+                    f"failed with exception:\n  {exc.__class__.__name__}: {exc!s}\n"
+                    "This may be due to a configuration or other issue with "
+                    "your system.\nFalling back to polling for file modifications.",
+                    fg="red",
+                    bold=True,
+                )
+
+    def stop(self):
+        self.observer.stop()
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, ex_type, ex_value, ex_tb):
+        self.stop()
+
+    def _start_observer(self, observer_class=Observer):
+        observer = observer_class()
+        for path in self.paths:
+            observer.schedule(self.event_handler, path, recursive=True)
+        observer.setDaemon(True)
+        observer.start()
+        self.observer = observer
 
     def is_interesting(self, time, event_type, path):
         # pylint: disable=no-self-use
         return True
 
     def __iter__(self):
-        if self.event_handler.queue is None:
-            raise RuntimeError("watcher used with callback")
+        # Alias this since we may need it during interpreter shutdown
+        queue_Empty = queue.Empty
+
         while 1:
             try:
                 item = self.event_handler.queue.get(timeout=1)
                 if self.is_interesting(*item):
                     yield item
-            except _Empty:
+            except queue_Empty:
```

Comment:
Maybe do a `from queue import Empty`, then we can do without the alias (and then the comment could be moved to the except, to ensure that it is not changed back)

Reply:
`from queue import Empty as QueueEmpty` maybe?

Reply:
> Maybe do a `from queue import Empty`, then we can do without the alias

Well (as you know) it's essentially the same thing.  My feeling is that using the assignment to create the reference to `queue.Empty` better indicates that we're doing it for a different reason than just to import it into the namespace.

> `from queue import Empty as QueueEmpty` maybe?

I like that better, though again, still, the purpose of what we're doing is not the same as a normal `import` (which is why @mitsuhiko wrote it as an assignment rather than import in the first place.)


My change just moves that from module to function scope, since we really only need to hold the reference in that one function.

Also note that, if this is converted to a function-scoped `import`, that would require a `# pylint: ignore=import-outside-toplevel`.

I'd rather leave it as is.  My second choice would be to put it back the way it was before I touched it.  Third choice would be a module-level `from queue import Empty as QueueEmpty` — but that would definitely require a comment with the `import` to warn as to why that is being done (rather than just referring to `queue.Empty`.)

Reply:
@yagebu @runfalk  Tell me if/how you want this adjusted.

(I was going to rebase to PR so that the CI lint test passes, but I'm holding off on that because I'm not sure whether these code comments will survive such a rebase.)

Reply:
My initial thought was to convert it to a module-level import, but you're right that having it in the function (as an assignment) is more explicit and "safer" from accidental changes in the future. So I'm also for leaving it as is now :)